pkg/client/kubernetes/cache/cache.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"strings"
+
+	"github.com/gardener/gardener/pkg/logger"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	runtimecache "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ cache.Cache = &Aggregator{}
+
+// Aggregator is a cache that can hold different cache implementations depending on the objects' GVKs.
+type Aggregator struct {
+	fallback   cache.Cache
+	gvkToCache map[schema.GroupVersionKind]cache.Cache
+	scheme     *runtime.Scheme
+}
+
+func (c *Aggregator) cacheForObject(obj runtime.Object) cache.Cache {
+	gvks, _, err := c.scheme.ObjectKinds(obj)
+	if err != nil || len(gvks) != 1 {
+		return c.fallback
+	}
+
+	return c.cacheForKind(gvks[0])
+}
+
+func (c *Aggregator) cacheForKind(kind schema.GroupVersionKind) cache.Cache {
+	gvk := kind
+	if strings.HasSuffix(gvk.Kind, "List") {
+		gvk.Kind = gvk.Kind[:len(gvk.Kind)-4]
+	}
+
+	cache, ok := c.gvkToCache[gvk]
+	if !ok {
+		cache = c.fallback
+	}
+	return cache
+}
+
+func (c *Aggregator) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	err := c.cacheForObject(obj).Get(ctx, key, obj)
+	if err != nil && !APIError(err) {
+		return NewCacheError(err)
+	}
```

Comment:
Can you elaborate why we need to distinguish between api errors and "cache errors"? 
What are the "cache errors" that could arise and that you want to handle differently?
And why don't we need such special handling in `IndexField`?

Reply:
The Multinamespace Cache returns an untyped error ([ref](https://github.com/kubernetes-sigs/controller-runtime/blob/4462fd4ebb88171af826622c1d6b73383aaf2cdf/pkg/cache/multi_namespace_cache.go#L132)) in case of a Get/List call for unknown namespaces. As this error is untyped, the idea was to ask for API errors and declare the rest as cache errors.

Reply:
Makes sense, thanks for clarifying.
It would nice to add an error type upstream, but this is obviously out of scope for this PR.
So all good for now.

Reply:
I added a comment to `pkg/client/kubernetes/client.go`